### PR TITLE
feat(date-input): add isReadOnly prop to date inputs and fields

### PR DESCRIPTION
**Pull request by jonnybel**
Relates to #695 

### Summary
This PR adds the `isReadOnly` prop to `DateInput`, `DateRangeInput`, and `DateTimeInput`, as well as their corresponding fields.

I also made a few adjustments to the calendar-body.styles in order to fix some inconsistencies with when a certain combination of states was active. The goal was to make it more consistent with other inputs.

#### Example 1:
Kapture 2019-11-27 at 16 43 34

#### Example 2:
Kapture 2019-11-27 at 16 47 54

### CODE REVIEW

File: src/components/inputs/date-input/date-input.js [hunk 1]
```diff
@@ -158,7 +158,6 @@ const DateInput = props => {
                     typeof props.placeholder === 'string'
                       ? props.placeholder
                       : intl.formatMessage(messages.placeholder),
-                  disabled: props.isDisabled,
```
Comment (jonnybel) [inline src/components/inputs/date-input/date-input.js]: We're already passing the `isDisabled` prop to CalendarBody. By also passing the `disabled` state here and directly to the input, it messes with the correct order of the states and the styles that should prevail.

File: src/components/inputs/date-input/date-input.js [hunk 2]
```diff
@@ -169,21 +168,22 @@ const DateInput = props => {
...
+                  onFocus: props.isReadOnly ? undefined : openMenu,
+                  onClick: props.isReadOnly ? undefined : openMenu,
                   ...filterDataAttributes(props),
                 })}
                 hasSelection={Boolean(selectedItem)}
                 onClear={clearSelection}
                 isOpen={isOpen}
                 isDisabled={props.isDisabled}
-                toggleButtonProps={getToggleButtonProps({
-                  disabled: props.isDisabled,
```
Comment (jonnybel) [inline src/components/inputs/date-input/date-input.js]: See comment for `inputProps`.

File: src/components/internals/calendar-body/calendar-body.js [hunk 3]
```diff
@@ -12,22 +12,18 @@ import {
 
 export const ClearSection = props => (
   <div
-    onClick={props.isDisabled ? undefined : props.onClear}
+    onClick={props.onClear}
     css={getClearSectionStyles(props)}
     aria-label="clear"
   >
-    {!props.isDisabled && (
```
Comment (jonnybel) [inline src/components/internals/calendar-body/calendar-body.js]: If we're already hiding the icon when `isDisabled`, then there's no point rendering the `ClearSection` at all. Which is why I removed all instances of the prop from here.

**Comment by montezume** [comment]
Can you make Percy run? Your PR wasn't open so it skipped it.

File: src/components/internals/calendar-body/calendar-body.js [hunk 4]
```diff
@@ -74,21 +70,26 @@ export const CalendarBody = props => {
...
           onBlur={handleInputBlur}
         />
-        {props.hasSelection && props.isClearable && (
+        {!disabledOrReadOnly && props.hasSelection && props.isClearable && (
           <ClearSection
-            isDisabled={props.isDisabled}
             hasError={props.hasError}
+            hasWarning={props.hasWarning}
             isFocused={isFocused}
+            isOpen={props.isOpen}
```
Comment (jonnybel) [inline src/components/internals/calendar-body/calendar-body.js]: These were missing.

File: src/components/internals/calendar-body/calendar-body.js [hunk 5]
```diff
@@ -127,11 +129,11 @@ CalendarBody.propTypes = {
   }),
   value: PropTypes.string,
   isDisabled: PropTypes.bool,
+  isReadOnly: PropTypes.bool,
   isOpen: PropTypes.bool,
   hasSelection: PropTypes.bool,
   hasWarning: PropTypes.bool,
   hasError: PropTypes.bool,
-  onClearPicker: PropTypes.func,
```
Comment (jonnybel) [inline src/components/internals/calendar-body/calendar-body.js]: This was probably a leftover.

File: src/components/internals/calendar-body/calendar-body.styles.js [hunk 6]
```diff
@@ -18,32 +18,30 @@ const getClearSectionStyles = props => {
     padding: ${vars.spacingXs};
     cursor: pointer;
     outline: 0;
+    transition: color ${vars.transitionStandard},
```
Comment (jonnybel) [inline src/components/internals/calendar-body/calendar-body.styles.js]: We have transitions in the `input`, so we should have in the other elements as well, or else they will change colours with different timings.

File: src/components/internals/calendar-body/calendar-body.styles.js [hunk 7]
```diff
@@ -18,32 +18,30 @@ const getClearSectionStyles = props => {
...
+      border-color ${vars.transitionStandard};
   `;
-  if (props.isDisabled) {
+  if (props.isOpen || props.isFocused) {
```
Comment (jonnybel) [inline src/components/internals/calendar-body/calendar-body.styles.js]: We're not rendering ClearSection when `isDisabled`.
And this other style that I added was missing.

File: src/components/internals/calendar-body/calendar-body.styles.js [hunk 8]
```diff
@@ -18,32 +18,30 @@ const getClearSectionStyles = props => {
...
-        border-color: ${vars.borderColorForInputWhenDisabled};
+        border-color: ${vars.borderColorForInputWhenFocused};
       `,
     ];
   }
   if (props.hasError) {
     return [
       baseIconStyles,
       css`
-        color: ${vars.fontColorForInputWhenError};
```
Comment (jonnybel) [inline src/components/internals/calendar-body/calendar-body.styles.js]: There's no text in ClearSection.

File: src/components/internals/calendar-body/calendar-body.styles.js [hunk 9]
```diff
@@ -18,32 +18,30 @@ const getClearSectionStyles = props => {
     padding: ${vars.spacingXs};
     cursor: pointer;
     outline: 0;
+    transition: color ${vars.transitionStandard},
+      border-color ${vars.transitionStandard};
   `;
-  if (props.isDisabled) {
```
Comment (jonnybel) [inline src/components/internals/calendar-body/calendar-body.styles.js]: We're not rendering ClearSection when isDisabled anymore.

File: src/components/internals/calendar-body/calendar-body.styles.js [hunk 10]
```diff
@@ -62,6 +60,12 @@ const getCalendarIconContainerStyles = (props, state) => {
     border-top-right-radius: ${vars.borderRadiusForInput};
     border-bottom-right-radius: ${vars.borderRadiusForInput};
     outline: 0;
+    transition: color ${vars.transitionStandard},
+      border-color ${vars.transitionStandard};
+    &:active,
```
Comment (jonnybel) [inline src/components/internals/calendar-body/calendar-body.styles.js]: Just to behave more similar to the input.

File: src/components/internals/calendar-body/calendar-body.styles.js [hunk 11]
```diff
@@ -120,37 +133,14 @@ const getDateTimeInputStyles = (props, state) => {
...
-        cursor: not-allowed;
-      }
-
-      &:disabled,
-      &:read-only {
-        background-color: ${vars.backgroundColorForInputWhenDisabled};
-        color: ${vars.fontColorForInputWhenDisabled};
-        border-color: ${vars.borderColorForInputWhenDisabled};
-        opacity: 1; /* fix for mobile safari */
-      }
```
Comment (jonnybel) [inline src/components/internals/calendar-body/calendar-body.styles.js]: These styles are already applied elsewhere, and having them here creates conflicts with the correct order of the input states.

**Comment by jonnybel** [comment]
I've checked the Percy diffs and they seem OK.
One of the diffs is a fix for the Example 2, the others are just new added specs.

File: src/components/inputs/date-input/date-input.js [hunk 12]
```diff
@@ -169,21 +168,22 @@ const DateInput = props => {
                       clearSelection();
                     }
                   },
-                  onFocus: openMenu,
-                  onClick: openMenu,
+                  // we only do this for readOnly because the input
+                  // doesn't ignore these events, unlike when its disabled
+                  onFocus: props.isReadOnly ? undefined : openMenu,
```
Comment (tdeekens) [inline src/components/inputs/date-input/date-input.js]: Nit: `!props.isReadonly && openMenu` if null works.

Comment (jonnybel) [reply src/components/inputs/date-input/date-input.js]: Yeah I did that at first, but it sets to `false` when `isReadOnly` which is not acceptable.

**Pull request merged**